File: Libraries/graphics/surface/surface.cpp
#include "surface.hpp"

void gfx::Surface::loadFromBuffer(const std::vector<unsigned char>& buffer, int w, int h) {
    width = w;
    height = h;
    
    if(width * height * 4 != buffer.size())
        throw BufferSizeError("Surface buffer size mismatch");
    
    data = buffer;
}

void gfx::Surface::createEmpty(int w, int h) {
    width = w;
    height = h;
    data = std::vector<unsigned char>(width * height * 4, 0);
}

int gfx::Surface::getIndex(int x, int y) const {
    if(x < 0 || x >= width || y < 0 || y >= height)
        throw PixelOutOfBoundsError("Pixel out of bounds");
    return ((height - y - 1) * width + x) * 4;
}

gfx::Color gfx::Surface::getPixel(int x, int y) const {
    Color result;
    int index = getIndex(x, y);
    result.r = data[index + 0];
    result.g = data[index + 1];
    result.b = data[index + 2];
    result.a = data[index + 3];
    return result;
}

void gfx::Surface::setPixel(int x, int y, gfx::Color color) {
    if(x < 0 || x >= width || y < 0 || y >= height)
        throw PixelOutOfBoundsError("Pixel out of bounds");
    
    int index = getIndex(x, y);
    data[index + 0] = color.r;
    data[index + 1] = color.g;
    data[index + 2] = color.b;
    data[index + 3] = color.a;
}

const std::vector<unsigned char>& gfx::Surface::getData() const {
    return data;
}

int gfx::Surface::getWidth() const {
    return width;
}

int gfx::Surface::getHeight() const {
    return height;
}

void gfx::Surface::draw(int x, int y, const Surface &surface, Color color) {
    for(int y_ = 0; y_ < surface.getHeight(); y_++)
        for(int x_ = 0; x_ < surface.getWidth(); x_++) {
            Color c = surface.getPixel(x_, y_);
            c.r *= (float)color.r / 255;
            c.g *= (float)color.g / 255;
            c.b *= (float)color.b / 255;
            c.a *= (float)color.a / 255;
            setPixel(x + x_, y + y_, c);
        }
}

File: libraries/graphics/surface.rs
use anyhow::{anyhow, bail, Result};
use serde_derive::{Deserialize, Serialize};

use crate::libraries::graphics as gfx;

use super::Color;
use crate::libraries::serialization;

/// An image stored in RAM.
#[derive(Serialize, Deserialize, Clone)]
pub struct Surface {
    pub(super) pixels: Vec<Color>,
    size: gfx::IntSize,
}

impl Surface {
    /// A surface of transparent pixels. The count is computed in `usize`, not `u32`: a wrapped
    /// product would hand out a surface smaller than it claims to be, which everything
    /// downstream trusts it is not.
    #[must_use]
    pub fn new(size: gfx::IntSize) -> Self {
        Self {
            pixels: std::vec![Color::new(0, 0, 0, 0); size.0 as usize * size.1 as usize],
            size,
        }
    }

    /// Serializes the surface through `libraries::serialization`, compressed with snap.
    pub fn serialize_to_bytes(&self) -> Result<Vec<u8>> {
        let mut buffer = Vec::new();
        serialization::serialize_into(&mut buffer, &self)?;
        Ok(snap::raw::Encoder::new().compress_vec(&buffer)?)
    }

    /// The inverse of `serialize_to_bytes`. **The pixel count is checked against the declared
    /// size**, because nothing downstream re-checks: `GpuDevice::create_texture` tells wgpu the
    /// texture is `size` big and hands it `pixels`, and a mismatch is a validation error, which
    /// is a panic. Surfaces come out of `.mod` files, which a player can replace.
    pub fn deserialize_from_bytes(buffer: &[u8]) -> Result<Self> {
        let decompressed = snap::raw::Decoder::new().decompress_vec(buffer)?;
        let surface: Self = serialization::deserialize(&decompressed)?;

        let expected = surface.size.0 as usize * surface.size.1 as usize;
        if surface.pixels.len() != expected {
            bail!("surface is {:?} but holds {} pixels rather than {expected}", surface.size, surface.pixels.len());
        }
        Ok(surface)
    }

    /// A 2D location as an index into the colour array.
    fn get_index(&self, pos: gfx::IntPos) -> Result<usize> {
        if pos.0 < 0 || pos.0 >= self.size.0 as i32 || pos.1 < 0 || pos.1 >= self.size.1 as i32 {
            bail!("Pixel out of bounds");
        }

        Ok((pos.1 * self.size.0 as i32 + pos.0) as usize)
    }

    pub fn get_pixel(&self, pos: gfx::IntPos) -> Result<&Color> {
        let index = self.get_index(pos)?;
        self.pixels.get(index).ok_or_else(|| anyhow!("Pixel array malformed"))
    }

    pub fn get_pixel_mut(&mut self, pos: gfx::IntPos) -> Result<&mut Color> {
        let index = self.get_index(pos)?;
        self.pixels.get_mut(index).ok_or_else(|| anyhow!("Pixel array malformed"))
    }

    #[must_use]
    pub const fn get_size(&self) -> gfx::IntSize {
        self.size
    }

    /// Copies another surface to the specified location, multiplied by `color`.
    pub fn draw(&mut self, pos: gfx::IntPos, surface: &Self, color: Color) -> Result<()> {
        let tint = |channel: u8, by: u8| (channel as f32 * (by as f32 / 255.0)) as u8;
        for (source_pos, source) in surface.iter() {
            *self.get_pixel_mut(pos + source_pos)? = Color::new(tint(source.r, color.r), tint(source.g, color.g), tint(source.b, color.b), tint(source.a, color.a));
        }

        Ok(())
    }

    /// Every pixel with its position, row by row. Walks the slice and derives the position
    /// from the index, which is the same order and skips a bounds check per pixel.
    pub fn iter(&self) -> impl Iterator<Item = (gfx::IntPos, &Color)> {
        let width = self.size.0 as i32;
        self.pixels.iter().enumerate().map(move |(index, pixel)| (index_to_pos(index, width), pixel))
    }

    /// Every pixel with its position, row by row, mutably.
    pub fn iter_mut(&mut self) -> impl Iterator<Item = (gfx::IntPos, &mut Color)> {
        let width = self.size.0 as i32;
        self.pixels.iter_mut().enumerate().map(move |(index, pixel)| (index_to_pos(index, width), pixel))
    }
}

/// The inverse of `get_index`. A zero width surface has no pixels, so the division is only
/// ever reached with a positive one.
const fn index_to_pos(index: usize, width: i32) -> gfx::IntPos {
    if width <= 0 {
        return gfx::IntPos(0, 0);
    }
    let index = index as i32;
    gfx::IntPos(index % width, index / width)
}
